### PR TITLE
Add Initial Support For Creation in Choices Inputs

### PR DESCRIPTION
**Pull request by djhi**
- [x] Add support for creation in `SelectInput`
- [x] Add support for creation in `AutocompleteInput`
- [x] Check support for creation in `ReferenceInput`
- [x] Add support for creation in `AutocompleteArrayInput`
- [x] Check support for creation in `ReferenceArrayInput`
- [x] Add support for creation in `SelectArrayInput`
- [x] Tests
- [x] Documentation

ra-quick-create

ra-quick-create-autocomplete-modal-reference

ra-quick-create-autocompletearray-modal-reference

### CODE REVIEW

File: packages/ra-core/src/dataProvider/useGetMany.ts [hunk 1]
```diff
@@ -99,13 +102,22 @@ const useGetMany = (
     const data = useSelector((state: ReduxState) =>
         selectMany(state, resource, ids)
     );
+    const version = useVersion(); // used to allow force reload
+    // used to force a refetch without relying on version
+    // which might trigger other queries as well
+    const [innerVersion, setInnerVersion] = useState(0);
```
[the remark this answers is not in the document]
Comment (djhi) [reply packages/ra-core/src/dataProvider/useGetMany.ts:108]: No good reason to use it actually

**Review by HALLERPierre — APPROVED** [review]

File: packages/ra-core/src/dataProvider/useGetMany.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (fzaninotto) [reply packages/ra-core/src/dataProvider/useGetMany.ts:108]: I agree with @HALLERPierre, you should use `useSafeSetState` to void calling `setInnerVersion` after the container component was unmounted.

File: packages/ra-core/src/form/useChoices.ts [hunk 2]
```diff
@@ -64,6 +65,12 @@ const useChoices = ({
                     ? optionText(choice)
                     : get(choice, optionText);
 
+            if (isValidElement<{ record: any }>(choiceName)) {
```
Comment (fzaninotto) [inline packages/ra-core/src/form/useChoices.ts:68]: Not sure why you added that. What's the use case?

File: packages/ra-ui-materialui/src/input/AutocompleteInput.spec.tsx [hunk 3]
```diff
@@ -684,19 +685,189 @@ describe('<AutocompleteInput />', () => {
...
+                validateOnBlur
+                onSubmit={jest.fn()}
+                render={() => <AutocompleteInput {...defaultProps} />}
+            />
+        );
+
+        expect(queryByRole('progressbar')).toBeNull();
+    });
+
+    test('should support creation of a new choice through the onCreate event', async () => {
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/input/AutocompleteInput.spec.tsx:698]: please use `it` as previous

File: packages/ra-ui-materialui/src/input/SelectArrayInput.spec.tsx [hunk 4]
```diff
@@ -303,17 +304,131 @@ describe('<SelectArrayInput />', () => {
...
-                    )}
+                    render={() => <SelectArrayInput {...defaultProps} />}
                 />
             );
 
             expect(queryByRole('progressbar')).toBeNull();
         });
     });
+
+    test('should support creation of a new choice through the onCreate event', async () => {
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/input/SelectArrayInput.spec.tsx:315]: please test with 'it'

File: packages/ra-ui-materialui/src/input/SelectInput.spec.tsx [hunk 5]
```diff
@@ -492,19 +493,140 @@ describe('<SelectInput />', () => {
...
+        await new Promise(resolve => setImmediate(resolve));
+        input.blur();
+
+        expect(
+            // The selector ensure we don't get the options from the menu but the select value
+            queryByText(newChoice.name, { selector: '[role=button]' })
+        ).not.toBeNull();
+    });
+
+    test('should support creation of a new choice through the onCreate event with a promise', async () => {
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/input/SelectInput.spec.tsx:544]: same

File: packages/ra-ui-materialui/src/input/useSupportCreateSuggestion.tsx [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+const CreateSuggestionContext = createContext<CreateSuggestionContextValue>(
+    undefined
+);
+
+interface CreateSuggestionContextValue {
+    filter?: string;
+    onCreate: (choice: any) => void;
+    onCancel: () => void;
+}
+export const useCreateSuggestion = () => useContext(CreateSuggestionContext);
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/input/useSupportCreateSuggestion.tsx:121]: bad name, as it does not actually create anything

**Review by fzaninotto — CHANGES_REQUESTED** [review]
Nice! I think we're missing an option in ReferenceInput that would handle the creation automatically. But let' do that in another PR.

File: packages/ra-ui-materialui/src/input/useSupportCreateSuggestion.tsx (hunk 6 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/input/useSupportCreateSuggestion.tsx:121]: You have a better one?

File: packages/ra-core/src/form/useChoices.ts (hunk 2 above)
Comment (djhi) [reply packages/ra-core/src/form/useChoices.ts:68]: Can't find the usecase anymore. I'll remove it

File: packages/ra-ui-materialui/src/input/useSupportCreateSuggestion.tsx (hunk 6 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/input/useSupportCreateSuggestion.tsx:121]: `useCreateSuggestionHelpers` ?

**Comment by djhi** [comment]
Messed up my commits/rebase big time :sob:

**Pull request closed without merging**